### PR TITLE
modules with basis/map coefficients

### LINKED ISSUE #18264: make map_coefficients in modules_with_basis safer or more general
<div id="comment:0"></div>

We would love to be able to, for example, present a symmetric function with factored coefficients.  But:

```
sage: f = SymmetricFunctions(ZZ).s()([2,2,1])
sage: f.map_coefficients(factor)
0
```

The result is not against the specification, because `map_coefficients` requires an endofunction.

There are two issues:

1.) we possibly want to be warned when supplying `map_coefficients` with something that is not an endofunction.  This might be achieved by adding an optional argument `check` with default `True`.

2.) we want some way to factor the coefficients.  Nicolas proposed to make something like

```
sage: f.map_coefficients(factor, codomain=SR)
```
or

```
sage: f.change_base_ring(SR).map_coefficients(factor) 
```
work.

Note that the `Polynomial` class contains something similar, namely:

```
def map_coefficients(self, f, new_base_ring=None)
```

However, there might be another tiny problem: it seems that currently it is not possible to make factored integers into a commutative ring.  This should possibly be a separate ticket.

CC:  @zabrocki @darijgr @sagetrac-sage-combinat @hivert

Component: **algebra**

Keywords: **map_coefficients, modules_with_basis, base_ring, check arguments**

Author: **Martin Rubey**

Branch/Commit: **[public/ticket/18264](https://github.com/sagemath/sagetrac-mirror/tree/public/ticket/18264) @ [`d0d80e7`](https://github.com/sagemath/sagetrac-mirror/commit/d0d80e73ccae7a5823e8692e72e0c31e59800e1b)**

Reviewer: **Nicolas M. Thiéry**

_Issue created by migration from https://trac.sagemath.org/ticket/18264_

### PR DESCRIPTION
**Pull request by mantepse**
This is a replacement for #37271, without dependency on the construction functor for symmetric functions.

Fixes #18264

### CODE REVIEW

File: src/sage/categories/modules_with_basis.py [hunk 1]
```diff
@@ -2062,17 +2063,25 @@ def trailing_term(self, *args, **kwds):
...
+            ``new_base_ring`` to override this behaviour.
+
+            An error is raised if the coefficients are not in the new base ring.
 
             INPUT:
 
-            - ``f`` -- an endofunction on the coefficient ring of the
-              free module
+            - ``f`` -- a callable that will be applied to the
+              coefficients of ``self``.
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]:
```suggestion
            - ``f`` -- a callable that will be applied to the
              coefficients of ``self``
```

File: src/sage/categories/modules_with_basis.py [hunk 2]
```diff
@@ -2062,17 +2063,25 @@ def trailing_term(self, *args, **kwds):
...
 
-            Return a new element of ``self.parent()`` obtained by applying the
-            function ``f`` to all of the coefficients of ``self``.
+            - ``new_base_ring`` (optional) -- if given, the resulting element
+              will be defined over this ring.
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]:
```suggestion
            - ``new_base_ring`` -- (optional) if given, the resulting element
              will be defined over this ring
```

File: src/sage/categories/modules_with_basis.py [hunk 3]
```diff
@@ -2062,17 +2063,25 @@ def trailing_term(self, *args, **kwds):
...
-            Mapping a function on coefficients.
+            Return the element obtained by applying ``f`` to the non-zero
+            coefficients of ``self``.
+
+            If ``f`` is a :class:`sage.categories.map.Map`, then the resulting
+            polynomial will be defined over the codomain of ``f``. Otherwise, the
+            resulting polynomial will be over the same ring as ``self``. Set
+            ``new_base_ring`` to override this behaviour.
+
+            An error is raised if the coefficients are not in the new base ring.
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]: As written, I don't think we want this error to be raised if the callable returns elements in, say `ZZ` but `new_base_ring=GF(2)`. We want to automatically convert the elements for consistency:
```
sage: R.<x,y> = ZZ[]
sage: x.map_coefficients(lambda p: 2*p, GF(2))
0
sage: x.map_coefficients(lambda p: GF(2)(p), QQ)
x
```
I think what you meant to say is if the coefficients must be able to be converted into `new_base_ring`.

Additionally, if `f` is a map, then `new_base_ring` should still be the ultimate target ring for consistency:
```
sage: R.<a> = QQ[]
sage: f = R.hom([a^2])
sage: S.<x,y> = R[]
sage: (a*x).map_coefficients(f)
a^2*x
sage: (a*x).map_coefficients(f, R['q'])
a^2*x
sage: _.parent()
Multivariate Polynomial Ring in x, y over Univariate Polynomial Ring in q over Univariate Polynomial Ring in a over Rational Field
```

File: src/sage/categories/modules_with_basis.py [hunk 4]
```diff
@@ -2096,8 +2105,32 @@ def map_coefficients(self, f):
...
+            R = self.parent()
+            if new_base_ring is not None:
+                B = new_base_ring
+                R = R.change_ring(B)
+            elif isinstance(f, Map):
+                B = f.codomain()
+                R = R.change_ring(B)
+            else:
+                B = self.base_ring()
+            return R.sum_of_terms((m, B(f(c))) for m, c in self)
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]:
```suggestion
            mc = self.monomial_coefficients(copy=False)
            return R.sum_of_terms((m, B(f(c))) for m, c in mc.items())
```
We should not assume specific `__iter__` output.

File: src/sage/categories/modules_with_basis.py [hunk 5]
```diff
@@ -2096,8 +2105,32 @@ def map_coefficients(self, f):
...
+            """
+            R = self.parent()
+            if new_base_ring is not None:
+                B = new_base_ring
+                R = R.change_ring(B)
+            elif isinstance(f, Map):
+                B = f.codomain()
+                R = R.change_ring(B)
+            else:
+                B = self.base_ring()
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]:
```suggestion
            if isinstance(f, Map):
                B = f.codomain()
            else:
                B = self.base_ring()
            if new_base_ring is not None:
                B = new_base_ring
            if B is not self.base_ring():
                R = R.change_ring(B)
```
To be consistent with what (multivariate) polynomials do.

**Review by tscrim — CHANGES_REQUESTED** [review]
Why is #37687 a dependency? This seems completely independent.

**Comment by mantepse** [comment]
> Why is #37687 a dependency? This seems completely independent.

The doctest only works with #37687.

**Comment by tscrim** [comment]
It would be better to use a doctest that does not depend on #37687. (In fact, this change is very mildly bad because it depends on something that is not implemented in the category.) Just use a regular CFM:
```
sage: M = CombinatorialFreeModule(ZZ, 'abc')
```

File: src/sage/categories/modules_with_basis.py (hunk 3 above)
Comment (mantepse) [reply src/sage/categories/modules_with_basis.py]: Do I understand correctly that you are asking me to adapt the documentation, the code is doing what it should, right?

**Comment by mantepse** [comment]
> It would be better to use a doctest that does not depend on #37687. (In fact, this change is very mildly bad because it depends on something that is not implemented in the category.) Just use a regular CFM:
> 
> ```
> sage: M = CombinatorialFreeModule(ZZ, 'abc')
> ```

OK

File: src/sage/categories/modules_with_basis.py (hunk 3 above)
Comment (tscrim) [reply src/sage/categories/modules_with_basis.py]: Yes, that's correct.

**Comment by tscrim** [comment]
Yes, that's what I was thinking. You don't need a new PR (it would make it *much* easier for me if you didn't); you can interactive rebase and just remove the commits from #37687 (among other ways) and force push the result.

**Comment by mantepse** [comment]
need more help with removing the offending commits, sorry.

**Comment by mantepse** [comment]
success!

File: src/sage/categories/modules_with_basis.py [hunk 6]
```diff
@@ -2062,17 +2063,26 @@ def trailing_term(self, *args, **kwds):
             """
             return self.parent().term(*self.trailing_item(*args, **kwds))
 
-        def map_coefficients(self, f):
+        def map_coefficients(self, f, new_base_ring=None):
             """
-            Mapping a function on coefficients.
+            Return the element obtained by applying ``f`` to the non-zero
+            coefficients of ``self``.
+
```
Comment (mkoeppe) [inline src/sage/categories/modules_with_basis.py:2071]: This method will need clarification because of:
- #37526

Comment (mantepse) [reply src/sage/categories/modules_with_basis.py:2071]: Could you please elaborate?  I don't see a change to `map_coefficients` in #37526, nor a deprecation.

Comment (mantepse) [reply src/sage/categories/modules_with_basis.py:2071]: ping?

Comment (mkoeppe) [reply src/sage/categories/modules_with_basis.py:2071]: Sorry for the delay. I was concerned regarding the meaning of "coefficients" when submodules are involved.
The code looks fine -- as it's just mapping monomial coefficients to monomial coefficients.

**Review by mkoeppe — APPROVED** [review]
LGTM

**Review by tscrim — APPROVED** [review]

**Pull request merged**